src/analyses/weaklyRelationalPointerAnalysis.ml

```diff
@@ -0,0 +1,197 @@
+
+  let body ctx f = ctx.local (*DONE*)
+
+  let assign_return ask t return_var expr =
+    (* the return value is not stored on the heap, therefore we don't need to remove any terms *)
+    match T.of_cil ask expr with
+    | (Some term, Some offset) -> meet_conjs_opt [Equal (return_var, term, offset)] t
+    | _ -> t
+
+  let return ctx exp_opt f =
+    let res = match exp_opt with
+      | Some e ->
+        assign_return (ask_of_ctx ctx) ctx.local (MayBeEqual.return_var (typeOf e)) e
+      | None -> ctx.local
+    in if M.tracing then M.trace "wrpointer-function" "RETURN: exp_opt: %a; state: %s; result: %s\n" d_exp (BatOption.default (MayBeEqual.dummy_lval (TVoid [])) exp_opt) (D.show ctx.local) (D.show res);res
+
+  let special ctx var_opt v exprs  =
+    let desc = LibraryFunctions.find v in
+    match desc.special exprs, v.vname with
+    | Assert { exp; refine; _ }, _ -> if not refine then
+        ctx.local
+      else
+        branch ctx exp true
+    | _, _ -> ctx.local
+
+  let duplicated_variable var = { var with vid = - var.vid - 4; vname = "wrpointer__" ^ var.vname ^ "'" }
+  let original_variable var = { var with vid = - (var.vid + 4); vname = String.lchop ~n:11 @@ String.rchop var.vname }
+
+  (*First all local variables of the function are duplicated (by negating their ID),
+    then we remember the value of each local variable at the beginning of the function
+    by using the analysis startState. This way we can infer the relations between the
+    local variables of the caller and the pointers that were modified by the function. *)
+  let enter ctx var_opt f args =
+    (* add duplicated variables, and set them equal to the original variables *)
+    let added_equalities = T.filter_valid_pointers (List.map (fun v -> CC.Equal (T.term_of_varinfo (duplicated_variable v), T.term_of_varinfo v, Z.zero)) f.sformals) in
+    let state_with_duplicated_vars = meet_conjs_opt added_equalities ctx.local in
+    if M.tracing then M.trace "wrpointer-function" "ENTER1: var_opt: %a; state: %s; state_with_duplicated_vars: %s\n" d_lval (BatOption.default (Var (MayBeEqual.dummy_varinfo (TVoid [])), NoOffset) var_opt) (D.show ctx.local) (D.show state_with_duplicated_vars);
+    (* remove callee vars that are not reachable and not global *)
+    let reachable_variables =
+      f.sformals @ f.slocals @ List.map duplicated_variable f.sformals @ reachable_from_args ctx args
+    in
+    let new_state = D.remove_terms_not_containing_variables reachable_variables state_with_duplicated_vars in
+    if M.tracing then M.trace "wrpointer-function" "ENTER2: result: %s\n" (D.show new_state);
+    [ctx.local, new_state]
+
+  (*ctx caller, t callee, ask callee, t_context_opt context vom callee -> C.t
+     expr funktionsaufruf*)
+  let combine_env ctx var_opt expr f exprs t_context_opt t (ask: Queries.ask) =
+    ctx.local
+
+  (*ctx.local is after combine_env, t callee*)
+  let combine_assign ctx var_opt expr f args t_context_opt t (ask: Queries.ask) =
```

Comment:
Can the assignment of a return lval (if there is one) not be expressed in terms of a normal assignment here? That would save some custom code...

Reply:
What do you mean exactly? The normal assignment just calls the function `assign_lval`, and in `combine_assign` I use the same function for the assignment of a return lval